source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1602 @@
+
+  Its underlying modular configuration system is written in Nix language, and uses packages from Nixpkgs.
+  The operating system environment and services it provides are configured with Nix language.
+
+[nix-manual]: https://nixos.org/manual/nix/stable
+[nixpkgs-manual]: https://nixos.org/manual/nixpkgs/stable/#preface
+[nixos-manual]: https://nixos.org/manual/nixos/stable/index.html#preface
+[home-manager]: https://github.com/nix-community/home-manager
+
+## Overview
+
+Using the Nix language in practice entails multiple things:
+
+- language: syntax and semantics
+- function libraries: `builtins` and `pkgs.lib`
+- developer tools: testing, debugging, linting and formatting, ...
+- generic build mechanisms: `stdenv`, trivial builders, ...
+- composition and configuration mechanisms: `override`, `overrideAttrs`, overlays, `callPackage`, ...
+- ecosystem-specific packaging mechanisms: `buildGoModule`, `buildPythonApplication`, ...
+- NixOS module system: `config`, `option`, ...
+
+**This guide only covers language syntax and semantics**, briefly discusses function libraries, and at the end will direct you to resources on the other components.
+
+### What will you learn?
+
+This guide should enable you to read typical Nix language code and understand its structure.
+
+It shows the most common and distingushing patterns in the Nix language:
+
+- [assigning names and accessing values](names-values)
+- declaring and calling [functions](functions)
+- [built-in and library functions](libraries)
+- [side effects](side-effects) to obtain build inputs and produce build results
+
+It *does not* explain all Nix language features in detail.
+See the [Nix manual][manual-language] for a full language reference.
+
+[manual-language]: https://nixos.org/manual/nix/stable/language/index.html
+
+### What do you need?
+
+- Familiarity with software development
+- Familiarity with Unix shell, to read command line examples <!-- TODO: link to yet-to-be instructions on "how to read command line examples" -->
+- [](./install-nix), to run the examples
+
+### How long does it take?
+
+- **No experience with functional programming: 2 hours**
+
+  Carefully work through all definitions and examples.
+
+  Make sure you understand them by reading all detailed explanations.
+
+- **Familiar with functional programming: 1 hour**
+
+  Learn syntax and semantics by examining definitions and examples, and read detailed explanations if in doubt.
+
+- **Proficient with functional programming: 30 minutes**
+
+  The Nix language is largely unsurprizing to experienced functional programmers.
+
+  Read through the guide to get accustomed to the syntax and language features.
+
+We recommend running all examples and playing with them to validate your assumptions and test what you have learned.
+
+### How to run the examples?
+
+- Every valid piece of Nix language code is a *Nix expression*.
+- Nix expressions can contain other Nix expressions, that is, they can be nested.
```

Comment:
That was indeed a fix after running it against beginners, clarifying "Nix file = Nix expression". I originally took that note from @tazjin's nix-1p, because I would have found that information very valuable early on. I can clearly remember wondering what kind of thing was in a Nix file, and someone telling me that it evaluates to exactly one thing would have helped a lot.

But, as you say, it made people wonder if expressions can be nested, and indeed they can, so that appeared like a reasonable addition.

So while I agree that it may be too much information at this point, I'm inclined to listen beginners here rather than experts.

Reply:
I'm not talking about "Evaluating a Nix expression produces a single value.", that makes a lot of sense and I can see how that helps beginners. But the fact that expressions can be nested feels out of place here. If we need to point out more explicitly that expressions can contain other expressions, this might be better further down somewhere.